File: v2/p5/src/brute_force/basic/Boj15650.java
package brute_force.basic;

// N과 M (2)
// 조합 (중복 허용 x, 오름차순)

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Boj15650 {
    static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static final StringBuffer sb = new StringBuffer();
    static StringTokenizer st;
    static int n, m;
    static int[] selected;

    private static void rec(int k) {
        if (k == m + 1) {
            for (int i = 1; i <= m; i++)
                sb.append(selected[i]).append(' ');
            sb.append('\n');
            return;
        }

        for (int i = selected[k - 1] + 1; i <= n; i++) {
            selected[k] = i;
            rec(k + 1);
            selected[k] = 0;
        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        selected = new int[m + 1];

        rec(1);
        System.out.print(sb);
    }
}

File: review/kotlin/src/main/kotlin/backtracking/boj_15650.kt
package backtracking

import java.util.StringTokenizer

private var N = 0
private var M = 0
private val arr = ArrayList<Int>()
private val sb = StringBuffer()

private fun solve(start: Int, depth: Int) {
    if (depth == M) {
        arr.forEach { sb.append(it).append(' ') }
        sb.append('\n')
        return
    }

    for (i in start..N) {
        if (i in arr) continue

        arr.add(i)
        solve(i, depth + 1)
        arr.remove(i)
    }
}

fun boj_15650() = with(System.`in`.bufferedReader()) {
    val st = StringTokenizer(readLine())
    N = st.nextToken().toInt()
    M = st.nextToken().toInt()

    solve(1, 0)
    print(sb)
}